### ISSUE TITLE
Understanding the Cost Volume

### ISSUE THREAD

**Issue by soulslicer** [JiaRenChang/PSMNet#167]
I refer to this section of the code:

```
        cost = Variable(torch.FloatTensor(refimg_fea.size()[0], refimg_fea.size()[1]*2, self.maxdisp/4,  refimg_fea.size()[2],  refimg_fea.size()[3]).zero_()).cuda()

        for i in range(self.maxdisp/4):
            if i > 0 :
             cost[:, :refimg_fea.size()[1], i, :,i:]   = refimg_fea[:,:,:,i:]
             cost[:, refimg_fea.size()[1]:, i, :,i:] = targetimg_fea[:,:,:,:-i]
            else:
             cost[:, :refimg_fea.size()[1], i, :,:]   = refimg_fea
             cost[:, refimg_fea.size()[1]:, i, :,:]   = targetimg_fea
        cost = cost.contiguous()
```

Cost volume is of size (64, 48, 96, 312) where its made of 32 [48,96,312] tensor for left image and 32 [48,96,312] tensor for right image. But the way it is filled with the feature vector is very confusing. I cannot understand this part
```

             cost[:, :refimg_fea.size()[1], i, :,i:]   = refimg_fea[:,:,:,i:]
             cost[:, refimg_fea.size()[1]:, i, :,i:] = targetimg_fea[:,:,:,:-i]
```

Why is it that for increasing disparity, you only fill a subset of the columns? What is the rationale for this?

**Comment by JiaRenChang** [comment]
@soulslicer 
This operation is to match the left and right features at  the corresponding disparity.

**Comment by soulslicer** [comment]
Yes, but the cost volume is containing repeating information from the feature maps. 

From what I see here, as the disparity increases, you fill less and less features for the left image, and more and more features from the right image in the columnwise direction.

I still cannot understand the rationale. Obviously, having larger disparity requires a larger observation of the features spatially, so it makes sense to increase it for the right image, but why decrease it for left?

**Comment by JiaRenChang** [comment]
We assume that there are 1x5 images need to match as below.
Left: ■▼◆★▲    ;     Righ: □▽◇☆△

According to our matching code, the cost volume will become
at disp 0:     |     at disp 1:   |     at disp 2:  
■▼◆★▲    |    0▼◆★▲    |    00◆★▲  
□▽◇☆△     |    0□▽◇☆     |    00□▽◇

The occlusion will be filled with zeros.
